add-docker

ilya-noize opened the pull request:
### Проект ShareIt. Техническое задание. Часть 4

**Ставим проблему**
Пользователей приложения ShareIt становится больше.
Но не всё идёт гладко: приложение работает медленнее, пользователи чаще жалуются, что их запросы подолгу остаются без ответа.

_Решение:_ оптимизировать загрузку путём создания микросервисов с помощью Docker.

_Задачи:_ 
 - Разбить приложение ShareIt на два — shareIt-server и shareIt-gateway. Они будут общаться друг с другом через REST. Вынести в shareIt-gateway всю логику валидации входных данных — кроме той, которая требует работы с БД.
 - Настроить запуск ShareIt через Docker. Приложения shareIt-server, shareIt-gateway и база данных PostgreSQL должны запускаться в отдельном Docker-контейнере каждый. Их взаимодействие должно быть настроено через Docker Compose.

ilya-noize committed:
Let's get started add docker.

ilya-noize committed:
gateway - userClient

ilya-noize committed:
gateway - userClient

ilya-noize committed:
gateway - itemClient

ilya-noize committed:
gateway - bookingClient

ilya-noize committed:
gateway - itemRequestClient

ilya-noize committed:
gateway - app.properties

ilya-noize committed:
docker-compose.yml

ilya-noize committed:
gateway - Dockerfile
fix errors

ilya-noize committed:
delete

ilya-noize committed:
server - repack

ilya-noize committed:
gateway - fix errors

ilya-noize committed:
fix MapStruct

ilya-noize committed:
properties modify

ilya-noize committed:
properties modify

ilya-noize committed:
fix conflict gateway-server

ilya-noize committed:
fix conflict gateway-server

ilya-noize committed:
Merge remote-tracking branch 'origin/add-docker' into add-docker

ilya-noize committed:
fix conflict gateway-server

ilya-noize committed:
fix controllers

ilya-noize committed:
fix gateway bugs

ilya-noize committed:
fix gateway bugs

ilya-noize committed:
fix gateway bugs

ilya-noize committed:
fix gateway bugs

ilya-noize committed:
Merge remote-tracking branch 'origin/add-docker' into add-docker

# Conflicts:
#	gateway/src/main/java/ru/practicum/shareit/item/api/controller/ItemController.java
#	gateway/src/main/resources/application.properties
#	server/src/main/resources/application.properties

ilya-noize committed:
Disabled Server tests

ilya-noize committed:
try start

ilya-noize committed:
try start all

ilya-noize committed:
try start TZ=Europe/Moscow

ilya-noize committed:
try start; fix long

ilya-noize committed:
Merge remote-tracking branch 'origin/add-docker' into add-docker

ilya-noize committed:
try start; fix db-connection

ilya-noize committed:
try start; fix exception - gateway;

ilya-noize committed:
try start;
gateway:
 ValidPageable;
 User.create - Constant in Mapping

ilya-noize committed:
try start;
gateway:
 create Booking: fix validation
 BookingController: mapping const path

ilya-noize committed:
try start;
gateway:
 fix BookingSimpleDto

ilya-noize committed:
try start;
gateway:
 mapping const path;
 log in BookingController

ilya-noize committed:
without TimeZone

ilya-noize committed:
fix getAll items

Dim1983 commented on gateway/src/main/java/ru/practicum/shareit/constants/Constants.java:
```diff
@@ -0,0 +1,65 @@
+
+    String HEADER_USER_ID = "X-Sharer-User-Id";
+
+    String CREATE_BOOKING = "/bookings";
```
С константами не борщи. Эндпоинты явно уникальны для контроллеров. Достаточно в классе контроллера как private static final объявить.

Dim1983 commented on docker-compose.yml:
```diff
@@ -0,0 +1,38 @@
+version: '3.3'
```
Как рекомендация попробовать использовать networks. Это позволяет обращаться к контейнерам по имени а не на localhost стучаться.

Dim1983 commented on docker-compose.yml:
```diff
@@ -0,0 +1,38 @@
+    depends_on:
+      - db
+    environment:
+      - SPRING_DATASOURCE_URL=jdbc:postgresql://db:5432/shareit
```
Все переменные обычно выносятся в файл .env и из него берутся. Чтобы например эти креды не светить в compose.

Dim1983 commented on docker-compose.yml:
```diff
@@ -0,0 +1,38 @@
+      - SHAREIT_SERVER_URL=http://server:9090
+#      - TZ=Europe/Moscow
+
+  server:
```
Так же обычно выставляют лимиты по cpu b памяти для контейнеров.

Dim1983 reviewed (commented):
Привет. Оставил комментарии.

ilya-noize committed:
private final Constants in controllers;

ilya-noize committed:
fix naming

ilya-noize committed:
docker-compose.yml:
  constants in .env;
  network bridge: shareit-network;
  gateway - cpu 10p.c, ram 128M;
  server - cru 20p.c, ram 512M;

ilya-noize committed:
docker-compose.yml:
  fix constants;

ilya-noize committed:
docker-compose.yml:
  constants in .env;
  gateway - cpu 10p.c, ram 128M;
  server - cru 20p.c, ram 512M;

ilya-noize committed:
docker-compose.yml:
  gateway - cpu 10p.c, ram 128M;
  server - cru 20p.c, ram 512M;

ilya-noize committed:
docker-compose.yml:
  gateway - cpu 10p.c, ram 128M;
  server - cru 20p.c, ram 512M;

ilya-noize committed:
docker-compose.yml:
  gateway - cpu 10p.c, ram 128M;
  server - cru 20p.c, ram 512M;

ilya-noize committed:
Merge remote-tracking branch 'origin/add-docker' into add-docker

ilya-noize committed:
-

ilya-noize committed:
-

ilya-noize committed:
Merge remote-tracking branch 'origin/add-docker' into add-docker

ilya-noize committed:
-no limit

ilya-noize committed:
-cpu 0.50

ilya-noize committed:
-env.const

ilya-noize committed:
-env.const yaml

ilya-noize committed:
-fix const networks

ilya-noize committed:
Merge remote-tracking branch 'origin/add-docker' into add-docker

# Conflicts:
#	docker-compose.yml

ilya-noize committed:
-rename env

ilya-noize committed:
-fix env

ilya-noize committed:
-fix env

Dim1983 reviewed (approved):
Привет. Принял.

ilya-noize committed:
-fix networks
